### PR TITLE
Begin AAA Dashboard Backend

### PR DESCRIPTION
**Pull request by emord**
This is a big PR but that's mostly because of the UCR data source definitions. I'm working on getting a test dataset up for automated tests, but just finished the first aggregation of data.

This is a fairly large departure from the CAS dashboard schema. I've decided this is the correct path forward to solve some real issues we've encountered. Some differences include:

* No similar tables to child_health_monthly or ccs_record_monthly. Instead we will have a Woman, ccs_record, and Child models. These models will store all the current information of the case and some key information used for date filters (such as ranges of time that a woman was pregnant) (not in this PR, but there will be similar tables to the rollup tables like agg_awc.
* Detailed historical information will be stored in array fields that include the changes a case property has seen over time. These should only be used for detailed beneficiary information or background tasks. This was inspired by a limitation of the CAS dashboard where we can only display the past 6 months of a child's measurements because of performance issues with querying a different table for every month
* No large aggregation step. Each aggregation will only update with information that has come in since the previous aggregation. Its very similar to https://www.citusdata.com/blog/2018/06/14/scalable-incremental-data-aggregation/ and I would suggest reading that blog post to understand the workflow that is being attempted.
* I have abstracted to a Woman model as a Woman in this program can flow between pregnant and "eligible couple" It felt easiest to have one data source for a person.

review commit by commit. There is a lot more iteration that will be going on with this work. And I'd like to keep the PRs small. This should have no affect to any other projects and as such should be ok to merge. 

Some more of the work includes:

* populate the woman, ccs_record and history tables
* create aggregate tables for the location levels
* populate a test data set to add to the automated tests
* change the static numbers that were introduced in https://github.com/dimagi/commcare-hq/pull/23110

Likely interested parties: @calellowitz @czue @snopoke 
buddies @nickpell @millerdev

### CODE REVIEW

File: custom/aaa/models.py [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+    household_case_id = models.TextField(null=True)
+    person_case_id = models.TextField(primary_key=True)
+
+    opened_on = models.DateField()
+    closed_on = models.DateField(null=True)
+    pregnant_ranges = ArrayField(
+        DateRangeField(),
+        help_text="The ranges in which a ccs_record has been opened and the baby has not been born",
+        null=True
+    )
```
Comment (emord) [inline custom/aaa/models.py]: This is not a pattern that we've followed before. I have verified that something similar to the following works:

`'2018-02-01'::date <@ any(array[daterange('2015-01-06','2015-02-10'), daterange('2017-10-05','2017-10-11')])`

There is not currently an accessible way to implement that in django, but I believe it should be feasible by writing a custom Django query expression

File: custom/aaa/models.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
+"""Contains all models for the AAA Convergence Dashboard.
+
+Conventions followed in this file:
+    * Fields ending in _ranges represent a field's history from forms that are useful for filtering.
+      This is only used for fields that are required to filter for specific ranges (usually a date range).
+    * Fields ending in _history represent a field's entire history from forms.
```
Comment (snopoke) [inline custom/aaa/models.py]: Do we know how big these could get? Could be useful to profile.

File: custom/aaa/models.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+from __future__ import absolute_import, unicode_literals
+
+from django.contrib.postgres.fields import ArrayField, DateRangeField
+from django.db import models
+
+from corehq.apps.userreports.models import StaticDataSourceConfiguration, get_datasource_config
+from corehq.apps.userreports.util import get_table_name
+
+
+class LocationDenormalizedModel(models.Model):
```
Comment (snopoke) [inline custom/aaa/models.py]: Not that I think it applies heavily to these models but just mentioning for possibly future work that optimising column order should be considered for large tables: https://blog.2ndquadrant.com/on-rocks-and-sand/ 

It might actually make more sense for the UCR tables.

File: custom/aaa/models.py (hunk 1 above)
Comment (snopoke) [reply custom/aaa/models.py]: Similar to above, if this is going to be used for filtering the table, have we done any profiling to see how it performs on large tables?

File: custom/aaa/models.py [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+    awc_id = models.TextField(null=True)
+
+    # AAA/MoH Hierarchy
+    taluka_id = models.TextField(null=True)
+    phc_id = models.TextField(null=True)
+    sc_id = models.TextField(null=True)
+    village_id = models.TextField(null=True)
+
+    class Meta(object):
+        abstract = True
```
Comment (snopoke) [inline custom/aaa/models.py]: :+1:

File: custom/aaa/models.py [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+
+    A Woman will likely have many pregnancies over her lifetime.
+    If a Woman becomes pregnant within two years of a child being born, she will have two ccs_record cases.
+    If a Woman has twins and those twins are registered in the program before birth,
+        she will have one ccs_record for this pregnancy.
+    If a Woman has twins and those twins are not registered in the program before birth,
+        she will have two ccs_records for this pregnancy
+    """
+
+    person_case_id = models.TextField()
```
Comment (snopoke) [inline custom/aaa/models.py]: should this be a foreign key?

File: custom/aaa/models.py [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+    # need to investigate that there is not an app workflow or error state
+    # where a person has multiple child_health cases
+    person_case_id = models.TextField(unique=True)
+    child_health_case_id = models.TextField(primary_key=True)
+
+    # For now, these should be the same between child_health and person case
+    # but going with child_health as there's a potential for person case to live
+    # on after the child_health case stops being tracked (into adolescent girl, or woman)
+    opened_on = models.DateField(help_text="child_health.opened_on")
+    closed_on = models.DateField(help_text="child_health.opened_on", null=True)
```
Comment (snopoke) [inline custom/aaa/models.py]: help text should be `closed_on`

File: custom/aaa/models.py [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+            SELECT
+                household_case_id,
+                doc_id,
+                dob,
+                sex,
+                migration_status
+            FROM "{person_cases_ucr_tablename}"
+            WHERE inserted_at >= %(window_start)s AND inserted_at < %(window_end)s
+        ) person
+        WHERE child.person_case_id = person.doc_id
```
Comment (snopoke) [inline custom/aaa/models.py]: Is there a case where the row may not exist in the table yet?

File: custom/aaa/tasks.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        aggregation_window_start=window_start,
+        aggregation_window_end=window_end,
+    )
+
+    # implement lock
+
+    agg_query, agg_params = method(domain, window_start, window_end)
+    with connections['aaa-data'].cursor() as cursor:
+        cursor.execute(agg_query, agg_params)
+    agg_info.end_time = datetime.utcnow()
```
Comment (snopoke) [inline custom/aaa/tasks.py]: This seems a little dangerous to me since it could result in missing a few rows that get inserted between the first date and now. Is it possible to have the query return the max `inserted_at` which could be used for the end time?

**Review by snopoke — COMMENTED** [review]
Are the UCRs going to be used for anything outside of aggregations? Just wondering if it could be possible to purge old data from the UCRs.

**Comment by snopoke** [comment]
+1 on the change in aggregation method!

File: custom/aaa/tasks.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+def update_table(domain, slug, method):
+    window_start = AggregationInformation.objects.filter(
+        step=slug, aggregation_window_end__isnull=False
+    ).order_by('-created_at').values_list('aggregation_window_end').first()
+    if window_start is None:
+        window_start = datetime(1900, 1, 1)
+    else:
+        window_start = window_start[0]
+
+    window_end = datetime.utcnow()
```
Comment (millerdev) [inline custom/aaa/tasks.py]: Looks like this is using date-based aggregation windows rather than sequence-id based with the intricate locking procedure described in [the citus article](https://www.citusdata.com/blog/2018/06/14/scalable-incremental-data-aggregation/). Did you decide against the approach outlined in that article, or are you using a variant that does not have all the same characteristics?

File: custom/aaa/models.py (hunk 1 above)
Comment (emord) [reply custom/aaa/models.py]: I have not yet done profiling, but did check that this can be used with a GiST index. Unfortunately those aren't natively supported in django until 2.0, so I'll need to do something a bit more custom in the meantime

File: custom/aaa/models.py (hunk 2 above)
Comment (emord) [reply custom/aaa/models.py]: I am not sure and it will be different for every indicator where the history is stored. For example things like weight and height of a child are measured every month, but an eligible couple's history of family planning methods are likely much smaller.

Another thing that makes this harder is that for large values, these are stored in TOAST tables and not the main table (similar to how large strings are stored). 

These history methods are only intended to be used in background aggregation tasks or in single beneficiary view tables, so there should not be any large user facing queries that need to use them.

File: custom/aaa/models.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+
+class LocationDenormalizedModel(models.Model):
+    """Abstract base model for containing all the location fields necessary for querying."""
+
+    # support multiple domains so QA is easier
+    domain = models.TextField(null=False)
+
+    # Shared location hierarchy
+    state_id = models.TextField(null=True)
+    district_id = models.TextField(null=True)
```
[the remark this answers is not in the document]
Comment (emord) [reply custom/aaa/models.py]: I'm intentionally using denormalization for now. I'm attempting to balance getting something QA-able before the pilot (march 1) vs supporting future CitusDB backend vs supporting future reports vs similarities to the current CAS dashboard so hopefully we can consolidate the custom warehouses and its easier to spin up on reports.

I'm not sure if I have made the right trade-off but I think it was easiest to do this for now and possibly in the future have the location hierarchy be a "reference" table that's on every node for efficient JOINs. It's possible that normalizing now with a foreign key to a location table would be more performant and easy to migrate, but with an impending Citus migration I'm less confident that I understand all the moving pieces there.

File: custom/aaa/tasks.py (hunk 9 above)
Comment (emord) [reply custom/aaa/tasks.py]: Yeah. I'm using dates because we insert the current time in inserted_at and we do not have a sequential ID. inserted_at is the only increasing ID we have for now. (the primary key is doc_id which is random uuid)

File: custom/aaa/models.py (hunk 5 above)
Comment (emord) [reply custom/aaa/models.py]: I've avoided foreign keys for now as I'm not clear on all the implications with Citus and our workflows there. I wanted to make this with a migration in mind and wanted to keep it simple for that.

File: custom/aaa/models.py (hunk 7 above)
Comment (emord) [reply custom/aaa/models.py]: There will be many  person cases that will not exist in the child table.  there does exist a case where:

1) person case is updated
2) aggregation window happens
3) child case updated
4) aggregation window happens
5) person case is not in this inserted_at window.

I think this could be solved in the background with a task that monitors the child table to find records that have these key  columns NULL, or perhaps a two step aggregation where we update person cases that have changed and also update records with NULL columns. I may change this for the pilot to do full data aggregations in the short term while this is more solved

File: custom/aaa/models.py (hunk 3 above)
Comment (emord) [reply custom/aaa/models.py]: Agreed that there's probably more low hanging fruit than this, but up to 20% space savings does seem worth exploring eventually.

I had heard of struct padding in C but never quite made the connection to SQL.

**Comment by emord** [comment]
> Are the UCRs going to be used for anything outside of aggregations?

Not at the moment. I haven't been asked for any mobile reports at this time. 

>Just wondering if it could be possible to purge old data from the UCRs.

Most of these are case data UCRs so we can't really purge them (or at least a large percentage of them) as they'll be updated fairly frequently.

For form UCRs time based partitioning makes a lot of sense to make sure that old data does not affect query performance. Purging the data is possible as we usually don't care about any data that's very old, but its possible that we would need to rebuild these if a new report was requested and we lost some of the raw information in the aggregation.

For example if they wanted a report about # of forms of each of the form types submitted over the life of the project (so like # BP forms, # EC forms, # hh forms, etc), but we only saved the total number of forms per month (AWC submitted 7 forms, etc) then we'd need to rebuild all forms. (I realize that for that example we'd have easier ways to do this, but it was the best one I could think of at the time)

File: custom/aaa/tasks.py (hunk 9 above)
Comment (millerdev) [reply custom/aaa/tasks.py]: From the article:

> There might still be in-progress writes to the events table that were assigned lower sequence values, but are not yet visible when the aggregation runs

Since this is not using the locking trick, I think the same applies to dates the way they're used here, which means some rows may be missed by the aggregation. Also I think small date differences between servers could cause issues.

Comment (emord) [reply custom/aaa/tasks.py]: Agreed about the small date differences between different dbs making this messy. Its possible that we could define some amount of drift that is expected and have the aggregation window end there. So the end date is always `now() - 1 minute`. That would solve this for machines with drift of up to one minute. Currently this is not an issue because it will all be on one server. It will be good to look into what sort of guarantees Citus can provide about synchronization of nodes (thought currently that utcnow does come from the celery task in python. Possible that we could use `now()` ad defined by postgres).

I am planning to use the locking trick as commented in line 28. From my understanding the workflow would be

1. Set window's end time
2. Lock the table for writes
3. Release lock.
    * all writes to the database that were inserted_at < time at point 1 are guaranteed to have been written.
4. Agg data that has been inserted up until point 1.

Comment (millerdev) [reply custom/aaa/tasks.py]: > I am planning to use the locking trick as commented in line 28.

Ah, I didn't realize that comment was a TODO item.

Using dates rather than a sequence still seems to me more complex and harder to reason about. `datetime.utcnow()` is not monotonically increasing (clock drift is real, and the NTP service adjusts system time routinely). See [SO answer](https://stackoverflow.com/a/3527632/10840) for some examples of how complex a subject this is. If I were doing this I would use a sequence as described in the article.

Comment (emord) [reply custom/aaa/tasks.py]: Yeah. That does make a lot of sense. We don't currently have a way to add sequences to a UCR data source, but I imagine it would not be difficult. I would need to think more about adding it to case UCRs since those are updated and not just inserted (like the 99% case for forms).

I have [removed the where clauses](https://github.com/dimagi/commcare-hq/pull/23243/commits/f702bd968604037b5b24655b89d85f17400f4a1b) for now while I think about this. This means that the aggregation_window_start and window_end are not used and each aggregation is a full aggregation. I have kept those database as they will still be useful concepts for the monthly location rollup tables that are coming next.

**Comment by emord** [comment]
@snopoke @millerdev  Thanks a ton for the feedback today.

I've added some more aggregations to populate the models I have now. There's a bit of repeated code that should be cleaned up (mostly in tasks.py and the SQL could be better generalized with helpers).

Some things that I will be looking into as the pilot goes on:

- Using sequence IDs instead of time (and adding sequence ids to UCR)
- General profiling both with query times and sizes
- Foreign keys & normalization and how that may impact changing from postgres to Citus.
- Curating a test dataset that is able to sufficiently test for the edge cases in the app.

For now I think that it makes sense to keep this base schema, and I will profile it with the feedback you guys have given me in mind which will inform next steps. I'll also keep up to date with Citus work and make changes based on that as needed. Let me know what you think of this approach

**Comment by snopoke** [comment]
+1 on the discussion. I think it will be very valuable to come up with an aggregation workflow that we can trust and that we can run often. A few more thoughts from me:

* we could potentially query the DB to get the window end knowing that any data which get's added after that point will have a lower date (not sure if that's 100% true if there are uncommitted transactions that could have a lower date).
  * Doing that on a sharded DB would be more challenging since you'd want to take the min of the dates from all shards

* Using sequence IDs definitely seems better but it only works if you can run the aggregations on the shard databases. In the example they are giving they are sharding on site_id and so never aggregating across sites which means that all aggregation queries run locally on shards. We would need to be able to guarantee that (at least the first level aggregation) can always happen on the local shard database. Something that isn't clear to me in their example is that the rollup table isn't distributed (maybe they just left that out of the example).

**Comment by millerdev** [comment]
+1 sounds good to me.

Responding to @snopoke's idea:
> could potentially query the DB to get the window end knowing that any data which get's added after that point will have a lower date

Assuming you meant "higher date" (more recent)—or did I misunderstand what you were saying? I don't think this will work because of things like clock drift and NTP time adjustments.

We might be able to make it work if we only ever aggregate up to some point sufficiently in the past that it will never be overlapped by drift or clock adjustments. Basically selecting a time range into which we can guarantee new data will never be written. However, I still think this is going to be more complicated than using a sequence.

**Comment by emord** [comment]
> * Using sequence IDs definitely seems better but it only works if you can run the aggregations on the shard databases

Is possible that we could have this aggregation loop over supervisor ids? Would need to think about how that affects the MoH hierarchy but it doesn't seem unreasonable at the moment. Also fits in with the concept of many, smaller aggregations

**Comment by millerdev** [comment]
> * Using sequence IDs definitely seems better but it only works if you can run the aggregations on the shard databases

I was under the impression that a sharded table could share a single sequence between all shards on citus. Is this not true?

**Comment by emord** [comment]
Offlined with Simon and Daniel. We're on the same page about using this as a base for future work and profiling to define our next steps and its clear that it wouldn't be immediately scalable tomorrow

File: corehq/sql_db/connections.py [hunk 11]
```diff
@@ -22,6 +22,7 @@
 ICDS_UCR_ENGINE_ID = 'icds-ucr'
 ICDS_UCR_NON_DASHBOARD_ENGINE_ID = 'icds-ucr-non-dashboard'
 ICDS_TEST_UCR_ENGINE_ID = 'icds-test-ucr'
+AAA_DB_ENGINE_ID = 'aaa-data'
```
Comment (czue) [inline corehq/sql_db/connections.py]: sorry this is a really dumb question but what is "AAA"?

File: custom/aaa/models.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (czue) [reply custom/aaa/models.py]: > possibly in the future have the location hierarchy be a "reference" table that's on every node for efficient JOINs

no data to back this up, but this sounds intuitively like a good option to me, and would be how a standard "star schema" dimensional data warehouse design would tackle the problem.

this is quite a lot of redundant storage assuming one of the low-level ids can deterministically generate the rest.

also probably obvious, but if a district's state changes, then you have to update a huge amount of row-level data to handle that whereas if location is a dimension then you just update it in relevant rows of the location table which is much smaller.

File: custom/aaa/models.py [hunk 12]
```diff
@@ -0,0 +1,420 @@
...
+
+Conventions followed in this file:
+    * Fields ending in _ranges represent a field's history from forms that are useful for filtering.
+      This is only used for fields that are required to filter for specific ranges (usually a date range).
+    * Fields ending in _history represent a field's entire history from forms.
+      They are saved with the format [[date, value], [date, value]].
+      They are not guaranteed to be ordered by date.
+      They are not intended to be queried on, their interpretation should happen in python.
+      These history fields are saved in a separate model to ensure that the main model stays
+      fast and small for filtering.
```
Comment (czue) [inline custom/aaa/models.py]: What's an example of how these `history` fields would be used?

**Review by czue — COMMENTED** [review]
mostly just reviewed for my own education

File: corehq/sql_db/connections.py (hunk 11 above)
Comment (emord) [reply corehq/sql_db/connections.py]: ASHA, AWW, ANM. Its the more official name of REACH and I didn't want to get into a situation where in 3 years someone asks me why I named this REACH (but I'm sure this won't prevent this)

File: custom/aaa/models.py (hunk 12 above)
Comment (emord) [reply custom/aaa/models.py]: From description:

> * Detailed historical information will be stored in array fields that include the changes a case property has seen over time. These should only be used for detailed beneficiary information or background tasks. This was inspired by a limitation of the CAS dashboard where we can only display the past 6 months of a child's measurements because of performance issues with querying a different table for every month

So instead of generating a query for 60 tables over a child's first 5 years, this should just query the array to return 60 values detailing how it has grown

File: custom/aaa/models.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (emord) [reply custom/aaa/models.py]: > also probably obvious, but if a district's state changes, then you have to update a huge amount of row-level data to handle that whereas if location is a dimension then you just update it in relevant rows of the location table which is much smaller.

FWIW This doesn't happen right now, but it could in the future. Before this does happen on a large scale for CAS there should be more speccing and understanding of what they actually want. There's a case to be made that we would not want the districts state to change for historical data as it would point to the geography that was relevant at that time. 

There would be similar questions for beneficiaries moving between AWC or ASHAs. Should we report only the current ASHA, or have the beneficiary report the different ones? 

Not that any of that is actionable right now, but I've heard the talks of this workflow being added for the past year or so, but no progress has been made on actually doing it.
